File: jdollarx/src/main/java/com/github/loyada/jdollarx/custommatchers/IsPresent.java
package com.github.loyada.jdollarx.custommatchers;

import com.github.loyada.jdollarx.InBrowser;
import com.github.loyada.jdollarx.Path;
import com.github.loyada.jdollarx.PathParsers;
import org.hamcrest.Description;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;
import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

import javax.xml.xpath.XPathExpressionException;

/**
 * Internal implementation.
 */
public class IsPresent {

    public IsPresent(){}

    public Matcher<Path> in(final InBrowser browser) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public String toString() {
                return "browser page contains to given Path";
            }

            @Override
            public void describeTo(final Description description) {
                description.appendText("browser page contains " + el.toString());
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(CustomMatchersUtil.wrap(el) + " is absent");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                return browser.isPresent(el);
            }
        };
    }


    public Matcher<Path> in(final Document document) {
        return new TypeSafeMatcher<Path>() {
            private Path el;

            @Override
            public String toString() {
                return "document contains to given path";
            }

            @Override
            public void describeTo(final Description description) {
                description.appendText("document contains " + el.toString());
            }

            @Override
            protected void describeMismatchSafely(final Path el, final
            Description mismatchDescription) {
                mismatchDescription.appendText(CustomMatchersUtil.wrap(el) + " is absent");
            }

            @Override
            protected boolean matchesSafely(final Path el) {
                this.el = el;
                final NodeList nodes;
                try {
                     nodes = PathParsers.findAllByPath(document, el);
                } catch (XPathExpressionException e) {
                    throw new RuntimeException("could not parse");
                }
                return nodes.getLength() > 0;
            }
        };
    }
}

File: dollarx/src/main/java/com/github/loyada/dollarx/custommatchers/scalatest/IsPresent.scala
package com.github.loyada.dollarx.custommatchers.scalatest

import com.github.loyada.dollarx.{PathParsers, Path, Browser}
import org.scalatest.matchers.{MatchResult, BeMatcher}
import org.w3c.dom.Document


class IsPresent {
  def in(browser: Browser): BeMatcher[Path] = new BeMatcher[Path]{
    def apply(left: Path) =
      getMatchResult(left, existsInBrowser(browser))
  }

  def in(doc: Document): BeMatcher[Path] = new BeMatcher[Path]{
    def apply(left: Path) =
      getMatchResult(left, existsInDocument(doc))
  }

  private def getMatchResult(path: Path, f: (Path => Boolean)) = MatchResult(
    f(path),
    path.toString + " is expected to be present, but is absent",
    path.toString + " is expected to be absent, but is present"
  )

  private def existsInBrowser(browser: Browser) = (path: Path) => browser.isPresent(path)
  private def existsInDocument(doc: Document) = (path: Path) =>  PathParsers.findAllByPath(doc, path).getLength > 0

}

